### PR TITLE
Decouple parse & resolve steps

### PR DESCRIPTION
**Pull request by xonixx**
This PR is prepared in context of #144.

It implements the part

> However, that would require separating out the resolver to a separate pass ... which is probably good to do anyway, but a bit of work.

Since the change is reasonably big I would add couple clarifications.

1. I tried to do only minimal work to get the task in question done, though I could not resist some (hopefully) minor refactorings and restructuring. 
2. It appears that resolving step needs some positions information, but not for all nodes. So at this point to keep the things simple and limited I wired the position only through very few of AST nodes. If need be we can equip all AST nodes with positions in generic way later but no need at this point.
3. I decided that we don't need to move ALL things that were in resolve.go to separate step. To me some things logically belong more to parsing than resolution. Example: `multiExprs` thing.
4. Since the parsing & resolving needed the same approach to errors with positions I generified this a bit by introducing the `PositionError` which is now panic-thrown internally by both parser & resolver. However I kept the external parser interface intact to return documented `ParseError`.
5. I checked the test suite - seems to work fine.

### CODE REVIEW

**Comment by benhoyt** [comment]
Excellent, thanks for this. I'll review it -- hopefully over the next week or so. Note that I just merged another change, so there's a small conflict in `parser.go` which needs fixing.

**Comment by xonixx** [comment]
> Note that I just merged another change, so there's a small conflict in parser.go which needs fixing.

Merged master.

**Comment by xonixx** [comment]
@benhoyt Any update?

**Comment by benhoyt** [comment]
Not yet, sorry. I glanced at it briefly, but am now traveling overseas. It
might be another week or two, but I will get to this.

**Comment by xonixx** [comment]
Sure no problem. Have a pleasant trip!

File: internal/resolver/resolve.go [hunk 1]
```diff
@@ -280,11 +354,11 @@ func (p *parser) resolveVars(prog *Program) {
...
 	prog.Scalars = make(map[string]int)
 	prog.Arrays = make(map[string]int)
-	for name, info := range p.varTypes[""] {
-		_, isFunc := p.functions[name]
+	for name, info := range r.varTypes[""] {
+		_, isFunc := r.functions[name]
 		if isFunc {
 			// Global var can't also be the name of a function
-			panic(p.errorf("global var %q can't also be a function", name))
+			panic(prog.EndPos.Errorf("global var %q can't also be a function", name))
```
Comment (benhoyt) [inline internal/resolver/resolve.go:361]: This is only the program's "end pos" because we don't have access to a better position at this point. Ideally it'd be the position of the global var. For now, I think we should just make it position 1,1 (that's at last as good as the program's end pos) and remove `Program.EndPos`.

File: lexer/lexer.go [hunk 2]
```diff
@@ -32,6 +33,32 @@ type Position struct {
 	Column int
 }
 
+func (pos Position) Add(lines int, cols int) Position {
```
Comment (benhoyt) [inline lexer/lexer.go:36]: I would rather we didn't add this `Add` method. It seems a bit weird to "add" to a position, and it's only used in one place, where I think it'd actually be preferable to make the error position the name itself, rather than just past the name.

File: internal/resolver/resolve.go [hunk 3]
```diff
@@ -1,16 +1,107 @@
...
+	case *ast.VarExpr:
+		r.recordVarRef(n)
+
+	case *ast.ArrayExpr:
+		r.recordArrayRef(n)
+
+	case *ast.UserCallExpr:
+		name := n.Name
+		if r.locals[name] {
+			panic(n.Pos.Add(0, len(name)).Errorf("can't call local variable %q as function", name))
```
Comment (benhoyt) [inline internal/resolver/resolve.go:91]: Let's make this error position just the position of the name, `n.Pos`. I think it's slightly better. It will mean updating a couple of error positions in tests.

File: lexer/lexer.go [hunk 4]
```diff
@@ -32,6 +33,32 @@ type Position struct {
...
+	return Position{
+		Line:   pos.Line + lines,
+		Column: pos.Column + cols,
+	}
+}
+
+type PositionError struct {
```
Comment (benhoyt) [inline lexer/lexer.go:43]: Hmm, I can see why you've added this (for use in both parser and resolver), but I don't love how this is a new exported type in the lexer. It's only used internally. I think we should add this in package `ast` as `ast.PositionError` instead.

Additionally, I don't love the `Errorf` method being on the position type; I'd prefer it were just a method or function in the parser and/or resolver (it's a tiny function, so it's fine if it's duplicated in both if needed). Or you could add `ast.PosErrorf(pos, format, args...)` if you want to share it.

File: parser/parser.go [hunk 5]
```diff
@@ -5,7 +5,7 @@
 package parser
 
 import (
-	"fmt"
+	"github.com/benhoyt/goawk/internal/resolver"
```
Comment (benhoyt) [inline parser/parser.go:8]: Let's move this down into a separate block (with a blank line in between stdlib imports and project imports).

File: parser/parser.go [hunk 6]
```diff
@@ -45,35 +36,51 @@ type ParserConfig struct {
...
-		// errors internally, and they're caught here. This
-		// significantly simplifies the recursive descent calls as
-		// we don't have to check errors everywhere.
+		// The parser and resolver use panic with a *PositionError to signal parsing
+		// errors internally, and they're caught here. This significantly simplifies
+		// the recursive descent calls as we don't have to check errors everywhere.
 		if r := recover(); r != nil {
 			// Convert to ParseError or re-panic
-			err = r.(*ParseError)
+			err = &ParseError{*r.(*PositionError)}
```
Comment (benhoyt) [inline parser/parser.go:60]: Per comments in `lexer`, this will now be something like:

```
// Convert to *ast.PositionError or re-panic
posErr := r.(*ast.PositionError)
err = &ParseError{Position: posError.Position, Message: posErr.Message}
```

File: parser/parser.go [hunk 7]
```diff
@@ -19,16 +19,7 @@ import (
...
-	Position Position
-	// Error message.
-	Message string
-}
-
-// Error returns a formatted version of the error, including the line
-// and column numbers.
-func (e *ParseError) Error() string {
-	return fmt.Sprintf("parse error at %d:%d: %s", e.Position.Line, e.Position.Column, e.Message)
+	PositionError
```
Comment (benhoyt) [inline parser/parser.go:22]: Because this will now be `ast.PositionError` (per comments in `lexer`), let's revert the change to these fields (mainly for the sake of API documentation), and just copy the fields over when we want to create the `ParseError`.

We'll need to add `ParseError.Error` back too.

File: parser/parser.go [hunk 8]
```diff
@@ -45,35 +36,51 @@ type ParserConfig struct {
...
 // ParseProgram parses an entire AWK program, returning the *Program
 // abstract syntax tree or a *ParseError on error. "config" describes
 // the parser configuration (and is allowed to be nil).
 func ParseProgram(src []byte, config *ParserConfig) (prog *Program, err error) {
 	defer func() {
-		// The parser uses panic with a *ParseError to signal parsing
-		// errors internally, and they're caught here. This
-		// significantly simplifies the recursive descent calls as
-		// we don't have to check errors everywhere.
+		// The parser and resolver use panic with a *PositionError to signal parsing
```
Comment (benhoyt) [inline parser/parser.go:55]: `... panic with an *ast.PositionError ...`

File: parser/parser.go [hunk 9]
```diff
@@ -1059,12 +1028,37 @@ func (p *parser) userCall(name string, pos Position) *ast.UserCallExpr {
...
 		i++
 	}
 	p.expect(RPAREN)
-	call := &ast.UserCallExpr{false, -1, name, args} // index is resolved later
-	p.recordUserCall(call, pos)
-	return call
+	return &ast.UserCallExpr{false, ast.WillBeResolvedLater, name, args, pos}
+}
+
+// Record a "multi expression" (comma-separated pseudo-expression
```
Comment (benhoyt) [inline parser/parser.go:1038]: Yeah, I agree it makes sense to keep the MultiExpr stuff in the parser -- thanks.

File: internal/resolver/resolve.go [hunk 10]
```diff
@@ -180,98 +253,99 @@ func (p *parser) getScope(name string) (ast.VarScope, string) {
...
+	scope, funcName := r.getScope(name)
 	if scope == ast.ScopeSpecial {
-		panic(p.errorf("can't use scalar %q as array", name))
+		panic(expr.Pos.Errorf("can't use scalar %q as array", name))
 	}
-	expr := &ast.ArrayExpr{scope, 0, name}
-	p.arrayRefs = append(p.arrayRefs, arrayRef{funcName, expr, pos})
-	info := p.varTypes[funcName][name]
+	expr.Scope = scope
+	//expr := &ast.ArrayExpr{scope, 0, name}
```
Comment (benhoyt) [inline internal/resolver/resolve.go:276]: Debugging comment left in?

File: parser/parser.go [hunk 11]
```diff
@@ -83,19 +90,14 @@ type Program struct {
...
 	// be exported in package "parser", otherwise these could live in
 	// "internal/ast".)
-	Begin     []ast.Stmts
-	Actions   []ast.Action
-	End       []ast.Stmts
-	Functions []ast.Function
-	Scalars   map[string]int
-	Arrays    map[string]int
-	Compiled  *compiler.Program
+	ast.ResolvedProgram
```
Comment (benhoyt) [inline parser/parser.go:93]: At first I was a little hesitant to change this (as `Program` is exported), but I think I've convinced myself it's okay, especially given the above comment. Plus, the fields are still accessible (via the embedded `ResolvedProgram` struct) if anyone really is using them (they shouldn't be).

**Review by benhoyt — CHANGES_REQUESTED** [review]
Thanks for your work here. I think it's really good work overall. I've left a bunch of comments, mostly minor and hopefully easy enough to address.

File: internal/ast/walk.go [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+// A Visitor's Visit method is invoked for each node encountered by Walk.
+// If the result visitor w is not nil, Walk visits each of the children
+// of node with the visitor w, followed by a call of w.Visit(nil).
+type Visitor interface {
+	Visit(node Node) (w Visitor)
+}
+
+// Helper functions for common node lists. They may be empty.
+
+func WalkExprList(v Visitor, list []Expr) {
```
[the remark this answers is not in the document]
Comment (xonixx) [reply internal/ast/walk.go:14]: This is true for `WalkExprList` but `WalkStmtList` is used in https://github.com/xonixx/goawk/blob/decouple_parse_resolve/internal/resolver/resolve.go#L77

File: parser/parser.go (hunk 11 above)
Comment (xonixx) [reply parser/parser.go:93]: Ok so I assume we don't need to change anything here. Resolving then without changes. Please let me know if you think otherwise.

File: internal/resolver/resolve.go [hunk 13]
```diff
@@ -1,16 +1,107 @@
...
+	r.resolveUserCalls(prog)
+	r.resolveVars(resolvedProg)
+
+	return resolvedProg
+}
+
+func (r *resolver) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+
+	case ast.Function:
```
[the remark this answers is not in the document]
Comment (xonixx) [reply internal/resolver/resolve.go]: Done. Applied the same for `Action`.

File: lexer/lexer.go (hunk 4 above)
Comment (xonixx) [reply lexer/lexer.go]: Yeah, I've added `ast.PosErrorf(pos, format, args...)`

File: parser/parser.go (hunk 5 above)
Comment (xonixx) [reply parser/parser.go]: Sure. Strange, that gofmt doesn't take care of this.

**Comment by xonixx** [comment]
Hey @benhoyt! Hopefully, I've resolved all items except for the one with `WalkExprList`/`WalkStmtList` - please see my comment there. What do you think?

File: internal/resolver/resolve.go [hunk 14]
```diff
@@ -60,41 +151,28 @@ type varRef struct {
...
-
-// Signal the end of a function
-func (p *parser) stopFunction() {
-	p.funcName = ""
-}
-
-// Add function by name with given index
-func (p *parser) addFunction(name string, index int) {
-	p.functions[name] = index
+func (r *resolver) initResolve(config *Config) {
```
Comment (benhoyt) [inline internal/resolver/resolve.go:163]: Nit: this could be named just `init` now that it's on the `resolver` type. In fact, what about making it a `newResolver(config *Config)` function instead?

File: parser/parser.go [hunk 15]
```diff
@@ -45,35 +46,55 @@ type ParserConfig struct {
...
-		// errors internally, and they're caught here. This
-		// significantly simplifies the recursive descent calls as
-		// we don't have to check errors everywhere.
+		// The parser and resolver use panic with an *ast.PositionError to signal parsing
+		// errors internally, and they're caught here. This significantly simplifies
+		// the recursive descent calls as we don't have to check errors everywhere.
 		if r := recover(); r != nil {
 			// Convert to ParseError or re-panic
-			err = r.(*ParseError)
+			posError := *r.(*ast.PositionError)
```
Comment (benhoyt) [inline parser/parser.go:70]: Nit: `// Convert to PositionError or re-panic`

**Review by benhoyt — APPROVED** [review]
Great! Let's just fix these few comment nits and I'll squash and merge this!

**Comment by xonixx** [comment]
Done.

**Comment by benhoyt** [comment]
Thanks, merged!
